vcd/resource_vcd_nsxv_firewall_rule.go

```diff
@@ -0,0 +1,884 @@
+		return fmt.Errorf("could not create firewall rule type for update: %s", err)
+	}
+
+	_, err = edgeGateway.UpdateNsxvFirewallRule(updateFirewallRule)
+	if err != nil {
+		return fmt.Errorf("unable to update firewall rule with ID %s: %s", d.Id(), err)
+	}
+
+	return resourceVcdNsxvFirewallRuleRead(d, meta)
+}
+
+func resourceVcdNsxvFirewallRuleRead(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
+	if err != nil {
+		return fmt.Errorf(errorUnableToFindEdgeGateway, err)
+	}
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	// Detect if this is data source or resource field and pick correct ID field
+	var isDatasource bool
+	id := d.Id()
+	if id == "" {
+		if ruleId, ok := d.GetOk("rule_id"); ok {
+			id = ruleId.(string)
+			isDatasource = true
+		}
+
+	}
+
+	readFirewallRule, err := edgeGateway.GetNsxvFirewallRuleById(id)
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("unable to find firewall rule with ID %s: %s", d.Id(), err)
+	}
+
+	err = setFirewallRuleData(d, readFirewallRule, edgeGateway, vdc)
+	if err != nil {
+		return err
+	}
+
+	if isDatasource {
+		d.SetId(readFirewallRule.ID)
+	}
+
+	return nil
+}
+
+func resourceVcdNsxvFirewallRuleDelete(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+	vcdClient.lockParentEdgeGtw(d)
+	defer vcdClient.unLockParentEdgeGtw(d)
+
+	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
+	if err != nil {
+		return fmt.Errorf(errorUnableToFindEdgeGateway, err)
+	}
+
+	err = edgeGateway.DeleteNsxvFirewallRuleById(d.Id())
+	if err != nil {
+		return fmt.Errorf("error deleting firewall rule with id %s: %s", d.Id(), err)
+	}
+
+	d.SetId("")
+	return nil
+}
+
+// resourceVcdNsxvFirewallRuleImport  is responsible for importing the resource.
+// The following steps happen as part of import
+// 1. The user supplies `terraform import _resource_name_ _the_id_string_` command
+// 2a. If the `_the_id_string_` contains a dot formatted path to resource as in the example below
+// it will try to import it. If it is found - the ID is set
+// 2b. If the `_the_id_string_` starts with `list@` and contains path to edge gateway similar to
+// `list@org.vdc.edge-gw` then the function lists all firewall rules and their IDs in that edge
+// gateway.
+// 2c. If the `_the_id_string_` contains a dot formatted path with the 4th element starting with
+// substring 'ui-id:3' and number after color - the import function will try to lookup real ID and
+// import the rule
+// 2d. If the `_the_id_string_` does not match format described neither in '2a', '2b', '2c' a
+// usage error message is printed
+//
+// Example resource name (_resource_name_): vcd_nsxv_firewall_rule.my-test-fw-rule
+// Example import path (_the_id_string_): org.vdc.edge-gw.existing-firewall-rule-id
```

Comment:
As this is an example, it would be more valuable to see some real-life representative number instead of `.existing-firewall-rule-id`

Reply:
Changed